Statistics/Histograms/Histogram.cs

```diff
@@ -32,127 +38,218 @@ internal abstract class Histogram : IHistogram
-            Mean = histogram.Mean;
-            Median = histogram.Median;
-            Variance = histogram.Variance;
-            Skewness = histogram.Skewness;
-            StandardDeviation = histogram.StandardDeviation;
-            Range = histogram.Range;
-            SampleSize = histogram.SampleSize;
-            State = histogram.State;
-            bool isConverged = true;
-            List<IMessage> msgs = new List<IMessage>(histogram.Messages);
-            foreach (var result in convergenceResults)
+            double sum = 0;
+            for (Int64 i = 0; i < BinCounts.Length; i++)
             {
-                if (!result.Passed) isConverged = false;
-                msgs.Add(result.TestMessage);
+                sum += BinCounts[i];
             }
-            Messages = msgs;
+            return Convert.ToInt32(sum);
         }
-        #endregion
+
+        internal IRange<double> GetRange(double min, double max)
+        {
+            var range = IRangeFactory.Factory(min, max, true, true, true, false);
+            return range;
+        }
+        internal double GetMean()
+        {           
+            double sum = 0;
+            double min = Min;
+                for (int i = 0; i < BinCounts.Length; i++)
+                {
+                    sum += (min + i * BinWidth + 0.5 * BinWidth) * BinCounts[i];
+                }
+            double mean = SampleSize > 0 ? sum / SampleSize : double.NaN;
+            return mean;
+        }
+
+
+        internal double GetMedian()
+        {
+            double median = InverseCDF(0.5);
+            return median;
+
+        }
+
+        internal double GetVariance()
+        {
+            double deviation = 0, deviation2 = 0;
+
+            for (int i = 0; i < BinCounts.Length; i++)
+            {
+                double midpoint = Min + i * BinWidth + 0.5 * BinWidth;
+
+                deviation = midpoint - Mean;
+                deviation2 += deviation * deviation;
+
+            }
+            double variance = SampleSize > 1 ? deviation2 / (SampleSize - 1) : 0;
+            return variance;
+        }
+
+        internal double GetSkewness()
+        {
+            double deviation = 0, deviation2 = 0, deviation3 = 0;
+
+            for (int i = 0; i < BinCounts.Length; i++)
+            {
+                double midpoint = Min + i * BinWidth + 0.5 * BinWidth;
+
+                deviation += midpoint - Mean;
+                deviation2 += deviation * deviation;
+                deviation3 += deviation2 * deviation;
+
+            }
+            double skewness = SampleSize > 2 ? deviation3 / SampleSize / Math.Pow(Variance, 3 / 2) : 0;
+            return skewness;
+        }
+
+
 
         #region Functions
-        public static IHistogram Fit(IData data, int nBins = 100) => IHistogramFactory.Factory(data, nBins);
-        public static IHistogram Fit(IHistogram histogram, IData sample)
+        public void AddObservationToHistogram(IData data)
```

Comment:
I would love to see this as implemented with a double instead of an Idata object.

Reply:
can be done at a later date with a refactor